Index: base/ca/shared/profiles/ca/caIPAserviceCert.cfg
===================================================================
--- base/ca/shared/profiles/ca/caIPAserviceCert.cfg	(revision 2124)
+++ base/ca/shared/profiles/ca/caIPAserviceCert.cfg	(revision 2125)
@@ -20,12 +20,12 @@
 policyset.serverCertSet.1.default.params.name=CN=$request.req_subject_name.cn$, OU=pki-ipa, O=IPA 
 policyset.serverCertSet.2.constraint.class_id=validityConstraintImpl
 policyset.serverCertSet.2.constraint.name=Validity Constraint
-policyset.serverCertSet.2.constraint.params.range=365
+policyset.serverCertSet.2.constraint.params.range=740
 policyset.serverCertSet.2.constraint.params.notBeforeCheck=false
 policyset.serverCertSet.2.constraint.params.notAfterCheck=false
 policyset.serverCertSet.2.default.class_id=validityDefaultImpl
 policyset.serverCertSet.2.default.name=Validity Default
-policyset.serverCertSet.2.default.params.range=180
+policyset.serverCertSet.2.default.params.range=731
 policyset.serverCertSet.2.default.params.startTime=0
 policyset.serverCertSet.3.constraint.class_id=keyConstraintImpl
 policyset.serverCertSet.3.constraint.name=Key Constraint
