Index: base/native-tools/src/tkstool/secutil.c
===================================================================
--- base/native-tools/src/tkstool/secutil.c	(revision 2162)
+++ base/native-tools/src/tkstool/secutil.c	(revision 2163)
@@ -3180,8 +3180,6 @@
         sprintf (SECUErrorBuf, "Certificate is not valid");
     else if (err == SEC_ERROR_EXTENSION_NOT_FOUND)
         sprintf (SECUErrorBuf, "Certificate extension was not found");
-    else if (err == SEC_ERROR_EXTENSION_VALUE_INVALID)
-        sprintf (SECUErrorBuf, "Certificate extension value invalid");
     else if (err == SEC_ERROR_CA_CERT_INVALID)
         sprintf (SECUErrorBuf, "Issuer certificate is invalid");
     else if (err == SEC_ERROR_CERT_USAGES_INVALID)
Index: base/native-tools/src/p7tool/secutil.c
===================================================================
--- base/native-tools/src/p7tool/secutil.c	(revision 2162)
+++ base/native-tools/src/p7tool/secutil.c	(revision 2163)
@@ -3183,8 +3183,6 @@
         sprintf (SECUErrorBuf, "Certificate is not valid");
     else if (err == SEC_ERROR_EXTENSION_NOT_FOUND)
         sprintf (SECUErrorBuf, "Certificate extension was not found");
-    else if (err == SEC_ERROR_EXTENSION_VALUE_INVALID)
-        sprintf (SECUErrorBuf, "Certificate extension value invalid");
     else if (err == SEC_ERROR_CA_CERT_INVALID)
         sprintf (SECUErrorBuf, "Issuer certificate is invalid");
     else if (err == SEC_ERROR_CERT_USAGES_INVALID)
