diff -ru base/common/scripts/functions /vakwetu/6.1/pki/base/common/scripts/functions
--- base/common/scripts/functions       2011-08-15 15:51:43.745919766 -0700
+++ /vakwetu/6.1/pki/base/common/scripts/functions      2011-08-15 14:07:29.058928170 -0700
@@ -693,12 +693,18 @@
     # Invoke the initscript for this instance
     case $PKI_SUBSYSTEM_TYPE in
         ca|kra|ocsp|tks)
-            /usr/bin/runcon -t pki_${PKI_SUBSYSTEM_TYPE}_script_t $PKI_INSTANCE_INITSCRIPT start
-            rv=$?
+            if [ -x /usr/sbin/selinuxenabled ] && /usr/sbin/selinuxenabled; then
+                /usr/bin/runcon -t pki_${PKI_SUBSYSTEM_TYPE}_script_t \
+                     $PKI_INSTANCE_INITSCRIPT start
+                rv=$?
+            else
+                $PKI_INSTANCE_INITSCRIPT start
+                rv=$?
+            fi
             ;;
         ra|tps)
-    $PKI_INSTANCE_INITSCRIPT start
-    rv=$?
+            $PKI_INSTANCE_INITSCRIPT start
+            rv=$?
             ;;
     esac

